ord_schema/dataset_templating.py

```diff
@@ -0,0 +1,139 @@
+flags.DEFINE_string(
+    'input_spreadsheet', None,
+    'Path to a spreadsheet file (with a header row) defining '
+    'values to replace placeholders in the input_template.')
+flags.DEFINE_string('output', None, 'Filename for output Dataset.')
+flags.DEFINE_boolean('validate', True, 'If True, validate Reaction protos.')
+
+
+def read_spreadsheet(file_name):
+    """Reads a spreadsheet file."""
+    _, suffix = os.path.splitext(file_name)
+    if suffix in ['xls', 'xlsx']:
+        return pd.read_excel(file_name, dtype=str)
+    return pd.read_csv(file_name, dtype=str)
+
+
+def _replace(string, substitutions):
+    pattern = re.compile('|'.join(map(re.escape, substitutions.keys())))
+    return pattern.sub(lambda match: substitutions[match.group(0)], string)
+
+
+def generate_dataset(template_string, df, validate=True):
+    """Generates a Dataset from a template reaction formatted as text according
+    to entries in a pd.Dataframe.
+
+    Inputs:
+        template_string: The contents of a Reaction pbtxt where placeholder
+            values to be replaced are defined between dollar sign. For example,
+            a SMILES identifier value could be "$product_smiles$". Spaces
+            are not allowed.
+        df: Pandas Dataframe where each row corresponds to one reaction and
+            column names match placeholders in the template_string.
+        validate: Optional Boolean controlling whether Reaction messages should
+            be validated as they are defined. Defaults to True.
+
+    Returns:
+        A Dataset message.
+
+    Raises:
+        ValueError: If there is no match for a placeholder string in df.
+        ValueError: If validate is True and there are validation errors when
+            validating an enumerated Reaction message.
+
+    """
+    placeholders = set(re.findall(r'\$\w+\$', template_string))
+    for placeholder in placeholders:
+        if placeholder not in df.columns:
+            # Allow "$my_placeholder$" to match "my_placeholder" in df.
+            if placeholder[1:-1] not in df.columns:
+                raise ValueError(f'Placeholder {placeholder} not found as a'
+                                 ' column in dataset spreadsheet')
+            df.rename(columns={placeholder[1:-1]: placeholder}, inplace=True)
+
+    reactions = []
+    for _, substitutions in df[placeholders].iterrows():
+        reaction_text = _replace(template_string, substitutions)
+        reaction = text_format.Parse(reaction_text, reaction_pb2.Reaction())
+        if validate:
+            errors = validations.validate_message(reaction,
+                                                  raise_on_error=False)
+            if errors:
+                raise ValueError(f'Enumerated Reaction is not valid: {errors}')
+        reactions.append(reaction)
+
+    return dataset_pb2.Dataset(reactions=reactions)
+
+
+def main(argv):
+    del argv  # Only used by app.run().
+
+    if not FLAGS.input_template:
```

Comment:
Use flags.mark_flags_as_required(['input_template', 'input_spreadsheet']) down before the call the app.run() instead.